test/test_pipeline.py

```diff
@@ -182,203 +204,320 @@ def test_to_code():
     )
 
 
-def test_PipelineCodeGen_order_components():
-    dependency_map = {"a": ["aa", "ab"], "aa": [], "ab": ["aba"], "aba": [], "b": ["a", "c"], "c": ["a"]}
-    ordered = _PipelineCodeGen._order_components(dependency_map=dependency_map)
-    assert ordered == ["aa", "aba", "ab", "a", "c", "b"]
+def test_PipelineCodeGen_is_component_order_invariant():
+    pipeline_config = {
+        "pipelines": [
+            {
+                "name": "Query",
+                "type": "Pipeline",
+                "nodes": [
+                    {"name": "EsRetriever", "inputs": ["Query"]},
+                    {"name": "EmbeddingRetriever", "inputs": ["Query"]},
+                    {"name": "JoinResults", "inputs": ["EsRetriever", "EmbeddingRetriever"]},
+                ],
+            }
+        ]
+    }
+
+    doc_store = {
+        "name": "ElasticsearchDocumentStore",
+        "type": "ElasticsearchDocumentStore",
+    }
+    es_retriever = {
+        "name": "EsRetriever",
+        "type": "ElasticsearchRetriever",
+        "params": {"document_store": "ElasticsearchDocumentStore"},
+    }
+    emb_retriever = {
+        "name": "EmbeddingRetriever",
+        "type": "EmbeddingRetriever",
+        "params": {
+            "document_store": "ElasticsearchDocumentStore",
+            "embedding_model": "sentence-transformers/all-MiniLM-L6-v2",
+        },
+    }
+    join_node = {
+        "name": "JoinResults",
+        "type": "JoinDocuments",
+    }
+
+    component_orders = [
+        [doc_store, es_retriever, emb_retriever, join_node],
+        [es_retriever, emb_retriever, join_node, doc_store],
+        [join_node, es_retriever, emb_retriever, doc_store],
+    ]
+
+    expected_code = (
+        "elasticsearch_document_store = ElasticsearchDocumentStore()\n"
+        "es_retriever = ElasticsearchRetriever(document_store=elasticsearch_document_store)\n"
+        'embedding_retriever = EmbeddingRetriever(document_store=elasticsearch_document_store, embedding_model="sentence-transformers/all-MiniLM-L6-v2")\n'
+        "join_results = JoinDocuments()\n"
+        "\n"
+        "p = Pipeline()\n"
+        'p.add_node(component=es_retriever, name="EsRetriever", inputs=["Query"])\n'
+        'p.add_node(component=embedding_retriever, name="EmbeddingRetriever", inputs=["Query"])\n'
+        'p.add_node(component=join_results, name="JoinResults", inputs=["EsRetriever", "EmbeddingRetriever"])'
+    )
+
+    for components in component_orders:
+        pipeline_config["components"] = components
+        code = _PipelineCodeGen.generate_code(
+            pipeline_config=pipeline_config, pipeline_variable_name="p", generate_imports=False
+        )
+        assert code == expected_code
 
 
 @pytest.mark.parametrize("input", ["\btest", " test", "#test", "+test", "\ttest", "\ntest", "test()"])
-def test_PipelineCodeGen_validate_user_input_invalid(input):
+def test_validate_user_input_invalid(input):
     with pytest.raises(ValueError):
```

Comment:
If you can, check that the error actually contains what you expect and it's not an unrelated `ValueError` coming from another part of the codebase.

I would do this here:
```python
 with pytest.raises(ValueError) as e:
     _validate_user_input(input)
    assert "\btest" in str(e)  # Just an example
```

Same for all the other validation tests below.